src/raster/mdarray.rs

```diff
@@ -162,6 +162,16 @@ impl<'a> MDArray<'a> {
         let rv = unsafe {
             let data_type = GDALMDArrayGetDataType(self.c_mdarray);
 
+            // HACK: this is too strict, but see https://github.com/georust/gdal/issues/290
+            if !self.datatype().class().is_numeric()
+                || self.datatype().numeric_datatype() != T::gdal_type()
+            {
+                return Err(GdalError::UnsupportedMdDataType {
+                    data_type: self.datatype().class(),
+                    method_name: "GDALMDArrayRead",
+                });
+            }
```

Comment:
Maybe we can have the call of `datatype` only once.
Now, you have one C call and two fn-calls that wrap the C call.
(we could move this out of the unsafe block, then)

Moreover, can you elaborate why this is too strict?

Reply:
It's too strict if it works like the legacy API. There you could read an `u16` band into an `u8` array, or the other way around, and GDAL would do a reasonable conversion (by saturating the values). This implementation prevents that.

The solution would be to pass the buffer data type, but I don't know how to get a `GDALExtendedDataTypeH` from a `GDALDataType`.